pkg/controller/podautoscaler/horizontal.go

```diff
@@ -502,6 +463,42 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 	return a.updateStatusIfNeeded(hpaStatusOriginal, hpa)
 }
 
+// normalizeDesiredReplicas takes the metrics desired replicas value and
+// normalizes it based on the appropriate conditions (i.e. < maxReplicas, >
+// minReplicas, etc...)
+func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.HorizontalPodAutoscaler, currentReplicas int32, desiredReplicas int32) int32 {
```

Comment:
Do you think that a method on `HorizontalController` is the best place to extract this logic? I also considered defining a separate struct (i.e. `DesiredReplicaNormalizer` containing `hpa` and `currentReplicas`, and then defining on it a single `normalize` method which would perform essentially the same function.

I personally think it might be nice to start splitting some of this logic out of `HorizontalController`, which is getting fairly large, but I didn't want to introduce a new pattern without first having a little discussion.

Additionally, I think a separate struct seems to better support the increased test coverage stated as a goal in the ticket. Otherwise, we'd have to make `normalizeDesiredReplicas` a public method to test it independently (which I'm not sure makes sense in given the visibility of other methods defined on the struct).

Reply:
Why would we need to make `normalizeDesiredReplicas` public just to test it? Unittests are in the same package, so they can test private methods, or am I missing something here?

Regarding a dedicated struct - personally I wouldn't do it, as it feels like a bit of an overkill for a single function with ~25 LOC.

Reply:
Re making `normalizeDesiredReplicas` public - I had thought that `horizontal_test.go` would be in the `podautoscaler_test` package. But it looks like its in `podautoscaler`, so you are correct that it will have access to `normalizeDesiredReplicas`. My understanding from working in other languages is that if possible, its best to avoid testing public methods. But it sounds like that's more of a "gray" area in Go/k8s. If you're comfortable with testing private methods, I think it is definitely the lightest weight way to add this functionality.

And agreed, that makes sense to me re whether a separate struct is necessary.